### DISCUSSION TITLE
Authenticating with Shopify session tokens

### DISCUSSION THREAD

**Discussion by dhmacs** [remix-run/remix#2422] [Q&A]
We're building a Shopify embedded app, where we can't use cookies (Shopify uses a mechanism to regenerate session tokens client side every few seconds - [read more](https://shopify.dev/apps/auth/oauth/session-tokens)). 

Shopify gives you an `authenticatedFetch` that automatically fetches the session token and passes it as a header, is there a way to use a custom fetch function with Remix API?
If not, how can we pass a custom header when  the form submits?

**Comment by edlaver** [comment]
Hi @dhmacs , I'm doing the same (building a Shopify embedded app with Remix).

Originally I started using something similar to this suggestion of overwriting the fetch function: https://github.com/remix-run/remix/discussions/2223#discussioncomment-2311081 from @kiliman  

I later found out there's a library that looks like it will assist with this: https://github.com/mlegenhausen/fetch-intercept

However, looks like it may have some issues around dealing with 401s: https://github.com/mlegenhausen/fetch-intercept/issues/60

In the end I ended up using a ServiceWorker instead to intercept the fetch requests and fetch them using the session token on the ServiceWorker side.

Seems to work okay, but has required some additional faffing around get the session token for the service worker (needs to request it from the client side using message passing), so using fetch-intercept may be simpler.

Would be nicer if we could just hook into or wrap the fetch function as used by Remix though.

I'll be publishing my Remix template for Shopify Apps soon, will update once done.  

Would love to know how you've gotten on and if you found another solution?

**Reply by dhmacs** [reply]
We've ended up with a slightly less convenient DX, where we listen to form submit, and append the token to the form data before we send the request to the api

**Reply by pjtf93** [reply]
Hi! i'm trying to build my first Shopify app with Remix and ran into this issue, did you find a solution for this?
